modules/stochastic_tools/src/distributions/KernelDensity1D.C

```diff
@@ -0,0 +1,180 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "KernelDensity1D.h"
+#include "math.h"
+#include "libmesh/utility.h"
+#include "DelimitedFileReader.h"
+#include "SystemBase.h"
+#include "Assembly.h"
+#include "Normal.h"
+#include "Uniform.h"
+
+registerMooseObject("StochasticToolsApp", KernelDensity1D);
+
+InputParameters
+KernelDensity1D::validParams()
+{
+  InputParameters params = Distribution::validParams();
+  MooseEnum bandwidth_rule("silverman standarddeviation userdefined");
+  MooseEnum kernel_function("gaussian uniform");
+  params.addClassDescription("KernelDensity1D distribution");
+  params.addRequiredParam<MooseEnum>(
+      "bandwidth_rule", bandwidth_rule, "Bandwidth rule for evaluating the bandwith.");
+  params.addRequiredParam<MooseEnum>(
+      "kernel_function",
+      kernel_function,
+      "Kernel function determines the shape of the underlying kernel for the kernel density.");
+  params.addRangeCheckedParam<Real>("bandwidth",
+                                    1.0,
+                                    "bandwidth > 0",
+                                    "Bandwidth controls the smoothness of the kernel density.");
+  params.addParam<std::vector<Real>>("data", "The data vector.");
+  params.addParam<FileName>("file_name", "Dummy.csv", "Name of the CSV file.");
+  return params;
+}
+
+KernelDensity1D::KernelDensity1D(const InputParameters & parameters)
+  : Distribution(parameters),
+    _bandwidth_rule(getParam<MooseEnum>("bandwidth_rule")),
+    _kernel_function(getParam<MooseEnum>("kernel_function")),
+    _bandwidth(getParam<Real>("bandwidth")),
+    _data(getParam<std::vector<Real>>("data")),
+    _file_name(getParam<FileName>("file_name"))
+{
+  bool data_set(isParamValid("data"));
+  bool file_name_set(parameters.isParamSetByUser("file_name"));
+
+  if (data_set && file_name_set)
+  {
+    paramError("data", "data and file_name both cannot be set at the same time.");
+  }
+  if (file_name_set)
+  {
+    MooseUtils::DelimitedFileReader reader(_file_name);
+    reader.setHeaderFlag(MooseUtils::DelimitedFileReader::HeaderFlag::OFF);
+    reader.read();
+    std::vector<std::vector<Real>> data1 = reader.getData();
+    _data = data1[0];
```

Comment:
This is how I would change this code:
```suggestion
  params.addParam<FileName>("file_name", "Name of the CSV file.");
  params.addParam<std::string>("file_column_name", "Name of column in csv file to use, by default first column is used.");
  return params;
}

KernelDensity1D::KernelDensity1D(const InputParameters & parameters)
  : Distribution(parameters),
    _bandwidth_rule(getParam<MooseEnum>("bandwidth_rule")),
    _kernel_function(getParam<MooseEnum>("kernel_function")),
    _bandwidth(getParam<Real>("bandwidth"))
{
  if (isParamValid("data") && isParamValid("file_name"))
    paramError("data", "data and file_name both cannot be set at the same time.");
  else if (isParamValid("file_name"))
  {
    MooseUtils::DelimitedFileReader reader(getParam<FileName>("file_name"));
    reader.read();
    if (isParamValid("file_column_name")
       _data = reader.getData(getParam<std::string>("file_column_name"));
    else
       _data = reader.getData(0);
  }
  else if (isParamValid("data"))
    _data = getParam<std::vector<Real>>("data");
  else
    mooseError("Either 'data' or 'file_name' parameters must be specified to represent input data");
```

I might have made a mistake here, so don't commit this directly. You also won't need the `_file_name` variable if you use this logic.

Reply:
@zachmprince This works perfectly! All the tests pass. Thanks.